watchfiles/main.py

```diff
@@ -191,39 +188,33 @@ async def stop_soon():
         stop_event_: 'AnyEvent' = anyio.Event()
     else:
         stop_event_ = stop_event
-    interrupted = False
-
-    async def signal_handler() -> None:
-        nonlocal interrupted
-
-        with anyio.open_signal_receiver(signal.SIGINT) as signals:
-            async for _ in signals:
-                interrupted = True
-                stop_event_.set()
-                break
 
     watcher = RustNotify([str(p) for p in paths], debug)
     timeout = _calc_async_timeout(rust_timeout)
+    CancelledError = anyio.get_cancelled_exc_class()
+
     while True:
         async with anyio.create_task_group() as tg:
-            # add_signal_handler is not implemented on Windows repeat ctrl+c should still stops the watcher
-            if exit_on_signal and sys.platform != 'win32':
-                tg.start_soon(signal_handler)
-            raw_changes = await anyio.to_thread.run_sync(watcher.watch, debounce, step, timeout, stop_event_)
+            try:
+                raw_changes = await anyio.to_thread.run_sync(watcher.watch, debounce, step, timeout, stop_event_)
+            except (CancelledError, KeyboardInterrupt):
+                stop_event_.set()
+                raise
             tg.cancel_scope.cancel()
 
         if raw_changes == 'timeout':
             if yield_on_timeout:
                 yield set()
             else:
                 logger.debug('rust notify timeout, continuing')
-        elif raw_changes == 'stop' or raw_changes == 'signal':
-            # cover both cases here although in theory the watch thread should never get a signal
-            if interrupted:
-                if raise_interrupt:
-                    raise KeyboardInterrupt
-                else:
-                    logger.warning('KeyboardInterrupt caught, stopping awatch')
+        elif raw_changes == 'stop':
+            return
+        elif raw_changes == 'signal':
+            # in theory the watch thread should never get a signal
+            if raise_interrupt:
+                raise KeyboardInterrupt
```

Comment:
I don't think this will work, since we raise the `KeyboardInterrupt` on line 202 above.

Reply:
This code should _never_ be hittable, as [`py.check_signals()`](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.check_signals) calls [`PyErr_CheckSignals`](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_CheckSignals) which always returns `0` if not on the main thread.

So I think `raise_interrupt==False` might make sense to _suppress_ a `KeyboardInterrupt` (and maybe a `CancelledError` but that would be somewhat surprising) after setting the `stop_event_` above. This just makes it a short-hand for `with contextlib.suppress` though, as it is in `watch`.

I also suspect that the equivalent call in `watch` is faulty for a different reason: if `PyErr_CheckSignals` returns `-1`, then the signal has been handled (and an exception raised and (I assume captured) and returned) and so raising `KeyboardInterrupt` would confuse a user who had installed a different `SIGINT` handler, or if the signal that raised the exception was not `SIGINT` at all. I guess that `return Ok("signal".to_object(py))` should probably return the value it got from `py.check_signals()` instead, and then a `KeyboardInterrupt` or whatever exception the installed signal handler specified would appear naturally as an exception from the `watcher.watch(debounce, step, rust_timeout, stop_event)` call.